promoted_python_delivery_client/client/two_arm_experiment.py

```diff
@@ -0,0 +1,66 @@
+from typing import Optional
+from promoted_python_delivery_client.model.cohort_arm import CohortArm
+from promoted_python_delivery_client.model.cohort_membership import CohortMembership
+
+
+class TwoArmExperiment:
+    def __init__(self,
+                 cohort_id: str,
+                 num_active_control_buckets: int,
+                 num_control_buckets: int,
+                 num_active_treatment_buckets: int,
+                 num_treatment_buckets: int) -> None:
+        if not cohort_id.strip():
+            raise ValueError("Cohort ID must be non-empty")
+
+        if num_control_buckets < 0:
+            raise ValueError("Control buckets must be positive")
+
+        if num_treatment_buckets < 0:
+            raise ValueError("Treatment buckets must be positive")
+
+        if num_active_control_buckets < 0 or num_active_control_buckets > num_control_buckets:
+            raise ValueError("Active control buckets must be between 0 and the total number of control buckets")
+
+        if num_active_treatment_buckets < 0 or num_active_treatment_buckets > num_treatment_buckets:
+            raise ValueError("Active treatment buckets must be between 0 and the total number of treatment buckets")
+
+        self.cohort_id = cohort_id
+        self.cohort_id_hash = hash(cohort_id)
+        self.num_active_control_buckets = num_active_control_buckets
+        self.num_control_buckets = num_control_buckets
+        self.num_active_treatment_buckets = num_active_treatment_buckets
+        self.num_treatment_buckets = num_treatment_buckets
+        self.num_total_buckets = num_treatment_buckets + num_control_buckets
+
+    def check_membership(self, user_id: str) -> Optional[CohortMembership]:
+        hash_code = _combine_hash(hash(user_id), self.cohort_id_hash)
+        bucket = abs(hash_code) % self.num_total_buckets
+        if bucket < self.num_active_control_buckets:
+            return CohortMembership(cohort_id=self.cohort_id, arm=CohortArm.CONTROL)
+
+        if (self.num_control_buckets <= bucket) and (bucket < (self.num_control_buckets + self.num_active_treatment_buckets)):
+            return CohortMembership(cohort_id=self.cohort_id, arm=CohortArm.TREATMENT)
+
+        return None
+
+
+def _combine_hash(first: int, second: int) -> int:
+    h = 17
+    h = h * 31 + first
+    h = h * 31 + second
```

Comment:
Side - for bucketing, it might make more sense to use bit xor to make the distribution of values more event.

Reply:
This implementation originated in the TypeScript client. I imagine there might be some advantage in having them consistent across all SDKs (for example, if multiple ones end up in use in the same environment).